### ISSUE TITLE
Binary Keys: incorrect node selection

### ISSUE THREAD

**Issue by AndrewBourgeois** [Netflix/dyno#156]
In file https://github.com/Netflix/dyno/blob/master/dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java

you have the following constructor:
```
private BaseKeyOperation(final byte[] k, final OpName o) {
        	this.key = null;
        	this.binaryKey = null;
        	this.op = o;
}
```
 which simply discards "k".

then in the Token Aware LB part (file https://github.com/Netflix/dyno/blob/master/dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/lb/TokenAwareSelection.java)
only the String keys are used to select a node: 
```
@Override
	public HostConnectionPool<CL> getPoolForOperation(BaseOperation<CL, ?> op) throws NoAvailableHostsException {
		
		String key = op.getKey();
		HostToken hToken = this.getTokenForKey(key);
		
		HostConnectionPool<CL> hostPool = null;
		if (hToken != null) {
			hostPool = tokenPools.get(hToken.getToken());
		}
		
		if (hostPool == null) {
			throw new NoAvailableHostsException("Could not find host connection pool for key: " + key + ", hash: " +
                    tokenMapper.hash(key));
		}
		
		return hostPool;
}
```

so it'll always return the first node when using binary keys which is kind of bad for performance.

**Comment by ipapapa** [comment]
For the first, you are correct. I will fix that. However, I am not following the second comment. The Token is derived from the key. What is the problem with that?

**Comment by AndrewBourgeois** [comment]
I'll rephrase the second comment:
fixing the BaseKeyOperation constructor by assigning "k" to "this.binaryKey" won't be enough to make binary keys work in Dyno.
[TokenAwareSelection](https://github.com/Netflix/dyno/blob/master/dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/lb/TokenAwareSelection.java) only uses BaseKeyOperation.getKey() and ignores the getBinaryKey() method.

**Comment by ipapapa** [comment]
@AndrewBourgeois sorry for delay on this.. Have been swamped with Dynomite and Dynomite-manager changes. Once complete, I will get to this.

**Comment by balajivenki** [comment]
@ipapapa  I am facing same issue and this might potentially cause 

https://github.com/Netflix/dynomite/issues/416

too. The DynoJedisClient.java is not honoring binary key so the routing is not happening as it should be.

One small thing that @AndrewBourgeois  missed in his earlier comment was that 

com.netflix.dyno.connectionpool.impl.lb.TokenAwareSelection#getPoolForOperation uses BaseOperation which doesnt have getBinaryKey. instead it should use BaseKeyOperation if we are actually using BaseKeyOperation.binaryKey. If we are going to still use the string value of byte[] as key then there is no need to change TokenAwareSelection and just set this.key as string value in BaseKeyOperation constructor.

**Comment by balajivenki** [comment]
@shailesh33  and @ipapapa  I just confirmed by modifying the dyno code at local base to change 

`private BaseKeyOperation(final byte[] k, final OpName o) {
        	this.key = null;
        	this.binaryKey = null;
        	this.op = o;
        }`

to 

`private BaseKeyOperation(final byte[] k, final OpName o) {
        	this.key = new String(k);
        	this.binaryKey = k;
        	this.op = o;
        }`

and this resolved the issue that is raised for https://github.com/Netflix/dynomite/issues/416

However I think dynomite might have to handle the case as well with the invalid routing.

The change is a hack to consider the binary key as its string value so the routing works as it is. This might now be the proper fix. This is one of a blocking issue for us. Can you please fix this for us and publish the new version to maven repository? Please.

**Comment by balajivenki** [comment]
@ipapapa  any update on this issue?

**Comment by shailesh33** [comment]
Hey, sorry no update on this. I was swamped with some prod issues past few
days. I will try to get to this by end of the week

On Thu, Feb 2, 2017 at 12:12 PM, Balaji Venkatesan <notifications@github.com

**Comment by ipapapa** [comment]
@balajivenki I have a couple of questions. 

1. Are you  getting a `NoAvailableHostsException` from Dyno prior to setting `byte[]` key as `String`? 
2. If an exception is not thrown from the Dyno client, why do you think we need to fix this in Dynomite? Dynomite should do the hashing of whatever it gets from the wire. Then Dynomite would do the hashing and re-routing properly. Isn't that the case?

**Comment by ipapapa** [comment]
Looking down in the implementation of the `HashPartitioner` we actually convert the String to a byte array. So it may be feasible to directly give the `byte[]`.

**Comment by balajivenki** [comment]
@ipapapa When we give byte[] the routing is not happening and it always goes to node1 in the local rack. Atlease that is what we get when we parse the response hostname from the response. This have caused some glitches in dynomite, but I do not experience that right now. This issue is purely on dyno client as the BaseKeyOperation constructor for binary key doesnt set the key and binarykey. So getKey() call gets as null and then routed to node1 all the time.

**Comment by ipapapa** [comment]
@balajivenki yes it will be sent to only one Dynomite node, but that node should properly reroute to whatever node is the owner of the that token. Right?

**Comment by balajivenki** [comment]
Yes @ipapapa  you are right, the issue we had was that since it is going to same node we had some glitches on dynomite side which I am not able to repro now and hence I joined the existing issue.

**Comment by shailesh33** [comment]
ok. I tried it locally (finally) and yes you will map to the same node because the hash will be 0.
But dynomite is doing the right thing and getting you the data that you need.
I think you can close the issue created on dynomite.

Talking about dynomite what were the glitches?

**Comment by balajivenki** [comment]
Sure I will @shailesh33 and @ipapapa 

The glitch was

`com.netflix.dyno.connectionpool.exception.FatalConnectionException: FatalConnectionException: [host=Host [hostname=UNKNOWN, ipAddress=UNKNOWN, port=0, rack: UNKNOWN, datacenter: UNKNOW, status: Down], latency=0(0), attempts=1]redis.clients.jedis.exceptions.JedisDataException: ERR Storage: Connection refused at com.netflix.dyno.jedis.JedisConnectionFactory$JedisConnection.execute(JedisConnectionFactory.java:104) at `

**Comment by ipapapa** [comment]
@balajivenki this looks a little more complicated to change in Dyno, but I am looking at it.

**Comment by ipapapa** [comment]
@balajivenki This seems to be relative big endeavor https://github.com/Netflix/dyno/compare/binaryKeys?expand=1. I have developed most of it and added new test cases that fail. Need to fix them. I am worried though If I have missed something structurally that would require even more work. Does the workaround with converting the string to byte array working for you? You can do it at the app side as well.
